README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# entrymodels: Estimate Entry Models

<!-- badges: start -->

[![CRAN
version](https://www.r-pkg.org/badges/version/entrymodels?color=orange)](https://cran.r-project.org/package=entrymodels)
[![CRAN
downloads](https://cranlogs.r-pkg.org/badges/entrymodels?color=blue)](https://cran.r-project.org/package=entrymodels)
[![Travis build
status](https://travis-ci.org/gnjardim/entrymodels.svg?branch=master)](https://travis-ci.org/gnjardim/entrymodels)
<!-- badges: end -->

Tools for measuring empirically the effects of entry in concentrated
markets, based in [Bresnahan and Reiss
(1991)](https://www.jstor.org/stable/2937655).

## Installation

You can install the released version of `entrymodels` from
[CRAN](https://CRAN.R-project.org/package=entrymodels) with:

``` r
install.packages("entrymodels")
```

And the development version from
[GitHub](https://github.com/gnjardim/entrymodels) with:

``` r
# install.packages("devtools")
devtools::install_github("gnjardim/entrymodels")
```

Which should return something similar to:

    * installing *source* package 'entrymodels' ...
    ** using staged installation
    ** R
    ** inst
    ** byte-compile and prepare package for lazy loading
    ** help
    *** installing help indices
      converting help for package 'entrymodels'
        finding HTML links ... done
        aux_matrix                              html  
        br1                                     html  
        br2                                     html  
        em_2var                                 html  
        em_basic                                html  
        load_example_data                       html  
    ** building package indices
    ** testing if installed package can be loaded from temporary location
    *** arch - i386
    *** arch - x64
    ** testing if installed package can be loaded from final location
    *** arch - i386
    *** arch - x64
    ** testing if installed package keeps a record of temporary installation path
    * DONE (entrymodels)

Please note that you should have
[Rtools](https://cran.r-project.org/bin/windows/Rtools/) installed.

## Examples

### Basic Model

This is a basic example which shows you how to estimate a basic entry
model with our sample data.

``` r
library(entrymodels)

tb <- load_example_data()
(em <- em_basic(tb, "Populacao", "n_agencias"))
#> # A tibble: 5 x 4
#>   n_competitors critical_values[,1] alpha[,1] gamma[,1]
#>           <int>               <dbl>     <dbl>     <dbl>
#> 1             1               5238.     0.832      7.12
#> 2             2              18961.     0.832      8.19
#> 3             3              42038.     0.825      8.79
#> 4             4              75638.     0.819      9.20
#> 5             5             162958.     0.809      9.71
```

### Two-Variable Model

This is a basic example which shows you how to estimate a two-variable
entry model with our sample data.

``` r
library(entrymodels)

tb <- load_example_data()
(em <- em_2var(tb, "Populacao", "RendaPerCapita", "n_agencias"))
#> # A tibble: 5 x 4
#>   n_competitors critical_values[,1] alpha[,1] gamma[,1]
#>           <int>               <dbl>     <dbl>     <dbl>
#> 1             1               5127.      1.06      17.8
#> 2             2              19265.      1.06      19.2
#> 3             3              44458.      1.05      20.0
#> 4             4              79959.      1.05      20.6
#> 5             5             169059.      1.03      20.9
```

## Citation

To cite package `entrymodels` in publications use:

> Guilherme N. Jardim (2020). entrymodels: Estimate Entry Models. R
> package version 0.2.1.
> <https://CRAN.R-project.org/package=entrymodels>

A BibTeX entry for LaTeX users is:

``` 
  @Manual{entrymodels,
    title = {entrymodels: Estimate Entry Models},
    author = {Guilherme {N. Jardim}},
    year = {2020},
    note = {R package version 0.2.1},
    url = {https://CRAN.R-project.org/package=entrymodels},
  }
```


NEWS.md
# entrymodels 0.2.1
* Added arguments check for easier error detections.

# entrymodels 0.2.0
* Added estimated parameters in returned tibbles.


R/br1.R
#' Build our optimization function
#'
#' @param params Parameters to construct function
#' @param n Number of observations in \code{data}
#' @param N_max An \code{integer} indicating the maximum number of competitors
#' @param l_params Length of parameters vector
#' @param A1 Auxiliary matrix A1
#' @param A2 Auxiliary matrix A2
#' @param S Size of the market
#' @param N Vector of zeros
#' @return The function to be optimized
#'
#' @import stats

br1 <- function(params, n, N_max, l_params, A1, A2, S, N){

    ml <- rep(0, n)

    alpha <- params[1:(N_max)]
    gamma <- params[(N_max + 1):l_params]

    alpha1 <- A1%*%alpha
    gamma1 <- A2%*%gamma

    Pi_bar <- S%*%matrix(alpha1, nrow = 1, byrow = TRUE) - rep(1, n)%*%matrix(gamma1, nrow = 1, byrow = TRUE)

    ml[N == 0] <- log(1 - stats::pnorm(Pi_bar[, 1][N == 0], mean = 0, sd = 1))

    for (j in 1:(N_max-1)){
        ml[N == j] <- log((stats::pnorm(Pi_bar[, j][N == j], mean = 0, sd = 1)) -
                          (stats::pnorm(Pi_bar[,(j + 1)][N == j], mean = 0, sd = 1)))
    }

    ml[N == N_max] = log(stats::pnorm(Pi_bar[, N_max][N == N_max], mean = 0, sd = 1))

    f = -sum(ml)

    return(f)

}


R/br2.R
#' Build our optimization function
#'
#' @param params Parameters to construct function
#' @param n Number of observations in \code{data}
#' @param N_max An \code{integer} indicating the maximum number of competitors
#' @param A1 Auxiliary matrix A1
#' @param A2 Auxiliary matrix A2
#' @param S1 First variable for size of the market
#' @param S2 Second variable for size of the market
#' @param N Vector of zeros
#' @return The function to be optimized
#'
#' @import stats

br2 <- function(params, n, N_max, A1, A2, S1, S2, N){

    ml <- rep(0, n)

    alpha <- params[1:(N_max)]
    gamma <- params[(N_max + 1):(2*N_max)]

    alpha1 <- A1%*%alpha
    gamma1 <- A2%*%gamma

    Pi_bar <- (S1 + params[2*N_max + 1]*S2)%*%matrix(alpha1, nrow = 1, byrow = TRUE) - rep(1, n)%*%matrix(gamma1, nrow = 1, byrow = TRUE)

    ml[N == 0] <- log(1 - stats::pnorm(Pi_bar[, 1][N == 0], mean = 0, sd = 1))

    for (j in 1:(N_max-1)){
        ml[N == j] <- log((stats::pnorm(Pi_bar[, j][N == j], mean = 0, sd = 1)) -
                          (stats::pnorm(Pi_bar[,(j + 1)][N == j], mean = 0, sd = 1)))
    }

    ml[N == N_max] = log(stats::pnorm(Pi_bar[, N_max][N == N_max], mean = 0, sd = 1))

    f = -sum(ml)

    return(f)

}


R/em_basic.R
#' Basic Entry Model
#' @description Estimate basic entry model with only one variable for the market size.
#'
#' @param data A \code{data.frame} object containing your data
#' @param Sm A string indicating the market size variable, present in \code{data}
#' @param y A string indicating the outcome variable, present in \code{data}
#' @param N_max An \code{integer} indicating the maximum number of competitors. Defaults to 5.
#' @param alpha0 A \code{vector} of type \code{numeric} and length \code{N_max} indicating the initial condition for alpha. Defaults to a vector of 0.1's.
#' @param gamma0 A \code{vector} of type \code{numeric} and length \code{N_max} indicating the initial condition for gamma. Defaults to a vector of 1's.
#' @return A tibble with critical market sizes and estimated parameters, as explained in Bresnahan and Reiss (1991)
#'
#' @import stats
#' @importFrom magrittr %>%
#' @importFrom dplyr as_tibble
#' @importFrom dplyr select
#' @importFrom dplyr tibble
#'
#' @examples
#' tb <- data.frame(Sm = 1:5, y = 1:5)
#'
#' # estimate default model
#' em_n5 <- em_basic(tb, "Sm", "y")
#'
#' # estimate model with 3 competitors only
#' em_n3 <- em_basic(tb, "Sm", "y", N_max = 3)
#'
#'
#' \dontrun{
#' # estimate model with different initial conditions
#' em_difc <- em_basic(tb, "Sm", "y", alpha0 = rep(0.2, 5), gamma0 = rep(1.1, 5))
#'
#' # estimate model with example data
#' tb <- load_example_data()
#' em <- em_basic(tb, "Populacao", "n_agencias")
#' }
#'
#' @references
#' Bresnahan, T. F., & Reiss, P. C. (1991). Entry and competition in concentrated markets. Journal of political economy, 99(5), 977-1009.
#'
#' @author Guilherme N. Jardim, Department of Economics, Pontifical Catholic University of Rio de Janeiro
#'
#' @export

em_basic <- function(data, Sm, y, N_max = 5,
                     alpha0 = rep(0.1, N_max),
                     gamma0 = rep(1, N_max)) {

    # check arguments ---------------------------------------------------------
    if (!is.character(Sm) | !is.character(y)) {
        stop("Arguments 'Sm1', 'Sm2' and 'y' must be strings.")
    }

    if (N_max < 1 | N_max%%1 != 0) {
        stop("'N_max' must be an integer larger than 0")
    }


    ### to tibble
    data <- dplyr::as_tibble(data)


    # parameters --------------------------------------------------------------
    n <- nrow(data)
    N <- rep(0, n)


    # build N vector ----------------------------------------------------------
    n_ag <- data %>% dplyr::select(y)

    for (i in 1:(N_max-1)){
        N[n_ag == i] <- i
    }

    N[n_ag >= N_max] <- N_max


    # build auxiliary ---------------------------------------------------------
    aux <- aux_matrix(data, y, N_max, n)
    A1  <- aux[[1]]
    A2  <- aux[[2]]


    # initial conditions ------------------------------------------------------
    params0 <- c(alpha0, gamma0)


    # size of market ----------------------------------------------------------
    S <- data %>%
         dplyr::select(Sm) %>%
         log() %>%
         as.matrix()


    # optimization ------------------------------------------------------------
    A <- 1*diag(2*N_max)
    b <- rep(0, 2*N_max)
    l_params <- length(params0)

    br.restricted1 <- stats::constrOptim(theta = params0, f = br1,
                                         grad = NULL, ui = A, ci = b,
                                         n = n, N_max = N_max,
                                         A1 = A1, A2 = A2,
                                         S = S, N = N,
                                         l_params = l_params
                                         )

    alpha_star_restricted1 <- A1%*%br.restricted1$par[1:N_max]
    gamma_star_restricted1 <- A2%*%br.restricted1$par[(N_max+1):(length(br.restricted1$par))]
    em_out <- dplyr::tibble(n_competitors   = 1:N_max,
                            critical_values = exp(gamma_star_restricted1/alpha_star_restricted1),
                            alpha           = alpha_star_restricted1,
                            gamma           = gamma_star_restricted1
                            )


}


R/aux_matrix.R
#' Build our auxiliary matrices to estimate entry models
#'
#' @param data A \code{data.frame} object containing your data
#' @param y A string indicating the outcome variable
#' @param N_max An \code{integer} indicating the maximum number of competitors
#' @param n Number of observations in \code{data}
#' @return A list of the auxiliary matrices
#'
#' @importFrom dplyr select
#' @importFrom magrittr %>%

aux_matrix <- function(data, y, N_max, n) {

    # auxiliary matrix A1
    A1 <- (-1)*matrix(1, nrow = N_max, ncol = N_max, byrow = TRUE)

    for (i in 1:N_max){
        A1[i,1] <- 1
    }

    for (i in 1:N_max){
        for (j in 1:N_max){
            if (i < j){
                A1[i, j] <- 0
            }
        }
    }

    # auxiliary matrix A2
    A2 <- matrix(1, nrow = N_max, ncol = N_max, byrow = TRUE)

    for (i in 1:N_max){
        for (j in 1:N_max){
            if (i < j){
                A2[i, j] <- 0
            }
        }
    }

    return(list(A1, A2))
}


R/em_2var.R
#' Two-Variable Entry Model
#' @description Estimate entry model with two variables for the market size.
#'
#' @param data A \code{data.frame} object containing your data
#' @param Sm1 A string indicating the main market size variable, present in \code{data}
#' @param Sm2 A string indicating the second market size variable, present in \code{data}
#' @param y A string indicating the outcome variable, present in \code{data}
#' @param N_max An \code{integer} indicating the maximum number of competitors. Defaults to 5.
#' @param alpha0 A \code{vector} of type \code{numeric} and length \code{N_max} indicating the initial condition for alpha. Defaults to a vector of 0.1's.
#' @param gamma0 A \code{vector} of type \code{numeric} and length \code{N_max} indicating the initial condition for gamma. Defaults to a vector of 1's.
#' @return A tibble with critical market sizes and estimated parameters, as explained in Bresnahan and Reiss (1991)
#'
#' @import stats
#' @importFrom magrittr %>%
#' @importFrom dplyr as_tibble
#' @importFrom dplyr select
#' @importFrom dplyr tibble
#'
#' @examples
#' tb <- data.frame(Sm1 = 1:5, Sm2 = 1:5, y = 1:5)
#'
#' # estimate default model
#' em_n5 <- em_2var(tb, "Sm1", "Sm2", "y")
#'
#' # estimate model with 3 competitors only
#' em_n3 <- em_2var(tb, "Sm1", "Sm2", "y", N_max = 3)
#'
#'
#' \dontrun{
#' # estimate model with different initial conditions
#' em_difc <- em_2var(tb, "Sm1", "Sm2", "y", alpha0 = rep(0.2, 5), gamma0 = rep(1.1, 5))
#'
#' # estimate model with example data
#' tb <- load_example_data()
#' em <- em_2var(tb, "Populacao", "RendaPerCapita", "n_agencias")
#' }
#'
#' @references
#' Bresnahan, T. F., & Reiss, P. C. (1991). Entry and competition in concentrated markets. Journal of political economy, 99(5), 977-1009.
#'
#' @author Guilherme N. Jardim, Department of Economics, Pontifical Catholic University of Rio de Janeiro
#'
#' @export

em_2var <- function(data, Sm1, Sm2, y, N_max = 5,
                    alpha0 = rep(0.1, N_max),
                    gamma0 = rep(1, N_max)) {


    # check arguments ---------------------------------------------------------
    if (!is.character(Sm1) | !is.character(Sm2) | !is.character(y)) {
        stop("Arguments 'Sm1', 'Sm2' and 'y' must be strings.")
    }

    if (N_max < 1 | N_max%%1 != 0) {
        stop("'N_max' must be an integer larger than 0")
    }


    ### to tibble
    data <- dplyr::as_tibble(data)


    # parameters --------------------------------------------------------------
    n <- nrow(data)
    N <- rep(0, n)


    # build N vector ----------------------------------------------------------
    n_ag <- data %>% dplyr::select(y)

    for (i in 1:(N_max-1)){
        N[n_ag == i] <- i
    }

    N[n_ag >= N_max] <- N_max


    # build auxiliary ---------------------------------------------------------
    aux <- aux_matrix(data, y, N_max, n)
    A1  <- aux[[1]]
    A2  <- aux[[2]]


    # initial conditions ------------------------------------------------------
    k <- 2
    params0 <- c(alpha0, gamma0)


    # size of market ----------------------------------------------------------
    S1 <- data %>%
          dplyr::select(Sm1) %>%
          log() %>%
          as.matrix()

    S2 <- data %>%
          dplyr::select(Sm2) %>%
          log() %>%
          as.matrix()


    # optimization ------------------------------------------------------------
    A <- 1*diag(2*N_max)
    b <- rep(0, 2*N_max)
    l_params <- length(params0)

    br.restricted1 <- stats::constrOptim(theta = params0, f = br1,
                                         grad = NULL, ui = A, ci = b,
                                         n = n, N_max = N_max,
                                         A1 = A1, A2 = A2,
                                         S = S1, N = N,
                                         l_params = l_params
                                         )


    # initial conditions for second optim -------------------------------------
    alpha0  <- br.restricted1$par[1:N_max]
    gamma0  <- br.restricted1$par[(N_max+1):(length(br.restricted1$par))]
    beta0   <- rep(0.0001, k - 1)
    params0 <- c(alpha0, gamma0, beta0)


    # main optimization -------------------------------------------------------
    A <- diag(2*N_max + 1)
    A[2*N_max + 1, 2*N_max + 1] <- 0

    b <- rep(0, 2*N_max + 1)
    b[2*N_max + 1] <- -0.00001

    br.restricted2 <- stats::constrOptim(theta = params0, f = br2,
                                         grad = NULL, ui = A, ci = b,
                                         n = n, N_max = N_max,
                                         A1 = A1, A2 = A2,
                                         S1 = S1, S2 = S2,
                                         N = N
                                         )

    alpha_star_restricted2 <- A1%*%br.restricted2$par[1:N_max]
    gamma_star_restricted2 <- A2%*%br.restricted2$par[(N_max+1):(2*N_max)]

    S_critical_restricted2 <- dplyr::tibble(n_competitors   = 1:N_max,
                                            critical_values = exp(gamma_star_restricted2/alpha_star_restricted2 -
                                                                  br.restricted2$par[2*N_max + 1]*mean(S2)),
                                            alpha           = alpha_star_restricted2,
                                            gamma           = gamma_star_restricted2
                                            )


}
